lib/tree.rb

```diff
@@ -16,40 +16,168 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(Log n)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    # raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
+
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, new_node)
+    return new_node if current.nil?
+
+    if new_node.key <= current.key
+      current.left = add_helper(current.left, new_node)
+    else
+      current.right = add_helper(current.right, new_node)
+    end
+
+    current
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def find(key)
```

Comment:
👍 , but the space complexity is O(log n) because of the call stack, assuming the tree is balanced.